samza-core/src/main/java/org/apache/samza/operators/impl/WindowOperatorImpl.java

```diff
@@ -18,26 +18,323 @@
+
+  private enum TriggerType { EARLY, DEFAULT, LATE }
+
+  public WindowOperatorImpl(WindowOperatorSpec<M, WK, WV> spec) {
     window = spec.getWindow();
   }
 
   @Override
   public void onNext(M message, MessageCollector collector, TaskCoordinator coordinator) {
+    WindowKey<K> storeKey =  getStoreKey(message);
+    BiFunction<M, WV, WV> foldFunction = window.getFoldFunction();
+    WV wv = store.get(storeKey);
+
+    if (wv == null) {
+      wv = window.getInitializer().get();
+    }
+
+    WV newVal = foldFunction.apply(message, wv);
+    store.put(storeKey, newVal);
+
+    if (window.getEarlyTrigger() != null) {
+      TriggerKey<K> triggerKey = new TriggerKey<>(TriggerType.EARLY, storeKey);
+      getOrCreateTriggerWrapper(triggerKey, window.getEarlyTrigger())
+          .onMessage(message, collector, coordinator);
+    }
+
+    if (window.getDefaultTrigger() != null) {
+      TriggerKey<K> triggerKey = new TriggerKey<>(TriggerType.DEFAULT, storeKey);
+      getOrCreateTriggerWrapper(triggerKey, window.getDefaultTrigger())
+          .onMessage(message, collector, coordinator);
+    }
+  }
+
+  @Override
+  public void onTimer(MessageCollector collector, TaskCoordinator coordinator) {
+    long now = System.currentTimeMillis();
+    TriggerTimerState state;
+    while ((state = pendingCallbacks.peek()) != null && state.getScheduleTimeMs() < now) {
+      pendingCallbacks.remove();
+      state.getCallback().run();
+
+      TriggerImplWrapper wrapper = triggers.get(state.triggerKey);
+      if (wrapper != null) {
+        wrapper.onTimer(collector, coordinator);
+      }
+    }
+  }
+
+  private WindowKey<K> getStoreKey(M message) {
+    Function<M, K> keyExtractor = window.getKeyExtractor();
+    K key = null;
+
+    if (keyExtractor != null) {
+      key = keyExtractor.apply(message);
+    }
+
+    String windowId = null;
+    if (window.getWindowType() == WindowType.TUMBLING) {
+      long triggerDurationMs = ((TimeTrigger<M>) window.getDefaultTrigger()).getDuration().toMillis();
+      final long now = System.currentTimeMillis();
+      Long windowBoundary = now - now % triggerDurationMs;
+      windowId = windowBoundary.toString();
+    }
+
+    return new WindowKey<>(key, windowId);
+  }
+
+  private TriggerImplWrapper getOrCreateTriggerWrapper(TriggerKey<K> triggerKey, Trigger<M> trigger) {
+    TriggerImplWrapper wrapper = triggers.get(triggerKey);
+    if (wrapper != null) {
+      return wrapper;
+    }
+
+    TriggerHandlerImpl triggerHandler = new TriggerHandlerImpl();
+    TriggerImpl<M> triggerImpl = TriggerImpls.createTriggerImpl(trigger);
+    TriggerContextImpl triggerContext = new TriggerContextImpl(triggerKey);
+    wrapper = new TriggerImplWrapper(triggerKey, triggerImpl, triggerContext, triggerHandler);
+    triggers.put(triggerKey, wrapper);
+
+    return wrapper;
+  }
+
+  private void handleTrigger(TriggerKey<K> triggerKey, MessageCollector collector, TaskCoordinator coordinator) {
```

Comment:
`handleTrigger()` is called when the trigger when the window determines that the trigger has fired. So, I can rename it as `onTriggerFired`.